### ISSUE TITLE
Connection to swebhdfs

### ISSUE THREAD

**Issue by IamGianluca** [mtth/hdfs#58]
The cluster I'm trying to connect to uses secure webhdfs, so I've set up the `~/.hdfscli.cfg` in the following way.

```
[global]
default.alias = dev

[dev.alias]
url = https://<namenode>:14000
user = <my.username>
```

I've tried to upload a file as shown in the quickstart document but got an error message complaining about the SSL certificate.

```
$ hdfscli upload --alias=dev empty.txt models/
ERROR   Unexpected exception.
Traceback (most recent call last):
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/connectionpool.py", line 595, in urlopen
    chunked=chunked)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/connectionpool.py", line 352, in _make_request
    self._validate_conn(conn)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/connectionpool.py", line 831, in _validate_conn
    conn.connect()
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/connection.py", line 289, in connect
    ssl_version=resolved_ssl_version)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/util/ssl_.py", line 308, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/lib/python3.5/ssl.py", line 377, in wrap_socket
    _context=self)
  File "/usr/lib/python3.5/ssl.py", line 752, in __init__
    self.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 988, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 633, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/adapters.py", line 423, in send
    timeout=timeout
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/connectionpool.py", line 621, in urlopen
    raise SSLError(e)
requests.packages.urllib3.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/config.py", line 195, in wrapper
    return func(*args, **kwargs)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/__main__.py", line 260, in main
    progress=progress,
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 453, in upload
    hdfs_path = self.resolve(hdfs_path)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 220, in resolve
    root = self._get_home_directory('/').json()['Path']
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 92, in api_handler
    **self.kwargs
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 178, in _request
    **kwargs
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/sessions.py", line 596, in send
    r = adapter.send(request, **kwargs)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/adapters.py", line 497, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)
```

Does hdfscli support secure webhdfs connections?

**Comment by mtth** [comment]
It does, it just looks like the certificate used by the cluster isn't trusted by your machine (see [this stackoverflow question](http://stackoverflow.com/a/12864892/1062617) for more context).

You can work around this in various ways; for example using this [`SecureClient`](https://hdfscli.readthedocs.io/en/latest/advanced.html#custom-client-support) and setting `verify` to `False` in your configuration:

```cfg
[global]
default.alias = dev
autoload.paths = <path.to.secure.client.py>

[dev.alias]
client = SecureClient
url = https://<namenode>:14000
user = <my.username>
verify = false
```

**Comment by IamGianluca** [comment]
Thank you for the reply @mtth ! I've tried to create a SecureClient class which inherits from `hdfs.Client`:

```
from hdfs import Client
from requests import Session


class SecureClient(Client):

    """A new client subclass for handling HTTPS connections.

    Args:
        url: URL to namenode.
        cert: Local certificate. See `requests` documentation for details
        on how to use this.
        verify: Whether to check the host's certificate.
        **kwargs: Keyword arguments passed to the default `Client`
        constructor.
    """
    def __init__(self, url, cert=None, verify=True, **kwargs):
        session = Session()
        if cert is not None and ',' in cert:
            sessions.cert = [path.strip() for path in cert.split(',')]
        else:
            session.cert = cert
        if isinstance(verify, str): # Python 2.
            verify = verify.lower() in ('true', 'yes', 'ok')
        session.verify = verify
        super(SecureClient, self).__init__(url, session=session, **kwargs)
```

I then call it using the following test script:

```
from model.secureclient import SecureClient


client = SecureClient("https://<namenode>:14000", verify=False)
content = client.content("./", strict=False)  # to be honest I don't have any idea of what folders are inside the directory 
print(content)
```

Unfortunately this throws an `hdfs.util.HdfsError: Authentication failure. Check your credential` when calling the `content` method:

```
$ python main2.py
/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/connectionpool.py:838: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/security.html
  InsecureRequestWarning)
Traceback (most recent call last):
  File "main2.py", line 6, in <module>
    content = client.content("./", strict=False)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 262, in content
    res = self._get_content_summary(hdfs_path, strict=strict)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 81, in api_handler
    quote(client.resolve(hdfs_path), '/= '),
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 220, in resolve
    root = self._get_home_directory('/').json()['Path']
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 92, in api_handler
    **self.kwargs
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 181, in _request
    return _on_error(response)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 37, in _on_error
    raise HdfsError('Authentication failure. Check your credentials.')
hdfs.util.HdfsError: Authentication failure. Check your credentials.
```

How should I provide the password? I've tried to add both a `passwd` and `password` parameter in the `~/.hdfscli.cfg` file but neither solved the issue.

**Comment by mtth** [comment]
We'd need to see the underlying error to understand what's causing this. Could you enable logging and run the script again? For example, you can add the following lines at the top of your script:

```python
import logging
logging.basicConfig(level=logging.DEBUG)
```

**Comment by IamGianluca** [comment]
This is log:

```
$ python main2.py
INFO:hdfs.client:Instantiated <SecureClient(url='https://asrvdsdevhadoop.cloudapp.net:14000')>.
INFO:hdfs.client:Fetching content summary for './'.
DEBUG:hdfs.client:Resolved path '/' to '/'.
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): asrvdsdevhadoop.cloudapp.net
/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/requests-2.11.1-py3.5.egg/requests/packages/urllib3/connectionpool.py:838: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/security.html
  InsecureRequestWarning)
DEBUG:requests.packages.urllib3.connectionpool:"GET /webhdfs/v1/?op=GETHOMEDIRECTORY HTTP/1.1" 401 997
Traceback (most recent call last):
  File "main2.py", line 10, in <module>
    content = client.content("./", strict=False)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 262, in content
    res = self._get_content_summary(hdfs_path, strict=strict)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 81, in api_handler
    quote(client.resolve(hdfs_path), '/= '),
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 220, in resolve
    root = self._get_home_directory('/').json()['Path']
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 92, in api_handler
    **self.kwargs
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 181, in _request
    return _on_error(response)
  File "/home/gianluca/.virtualenvs/search/lib/python3.5/site-packages/hdfs/client.py", line 37, in _on_error
    raise HdfsError('Authentication failure. Check your credentials.')
hdfs.util.HdfsError: Authentication failure. Check your credentials.
```

**Comment by mtth** [comment]
Thanks, this didn't have as much information as I'd hoped unfortunately. I added an additional log statement in the latest release (`2.0.12`), would you mind upgrading and running the same command?

**Comment by sansagara** [comment]
Nice logging addition there!. It helped me come up with a Kerberos error i kept getting for Active Directory accounts, on my Kerberized cluster. I just opened an issue on requests-kerberos. FYI: https://github.com/requests/requests-kerberos/issues/82

**Comment by mtth** [comment]
I'm assuming you were able to address this, feel free to reopen if not.

**Comment by bmasna** [comment]
Hi, I have similar error connecting to hdfs in kerberos cluster. I enabled DEBUG and error looks as below.
Code:
from hdfs.ext.kerberos import KerberosClient
client = KerberosClient("http://{nnhost}:50070")
client.list("/tmp")
Error log:
DEBUG:requests_kerberos.kerberos_:handle_401(): Handling: 401
ERROR:requests_kerberos.kerberos_:generate_request_header(): authGSSClientStep() failed:
Traceback (most recent call last):
  File "/opt/anaconda3-5.0.1/lib/python3.6/site-packages/requests_kerberos/kerberos_.py", line 230, in generate_request_header
    negotiate_resp_value)
kerberos.GSSError: (('Unspecified GSS failure.  Minor code may provide more information', 851968), ('No Kerberos credentials available (default cache: /tmp/krb5cc_21000)', -1765328243))
ERROR:requests_kerberos.kerberos_:(('Unspecified GSS failure.  Minor code may provide more information', 851968), ('No Kerberos credentials available (default cache: /tmp/krb5cc_21000)', -1765328243))
Traceback (most recent call last):
  File "/opt/anaconda3-5.0.1/lib/python3.6/site-packages/requests_kerberos/kerberos_.py", line 230, in generate_request_header
    negotiate_resp_value)
Can you suggest how to pass kerberos credentials to python?

**Comment by tritab** [comment]
@bmasna your question is probably better suited to be a new issue. However, did you try to kinit before making the request?
